### DISCUSSION TITLE
Selectable text

### DISCUSSION THREAD

**Discussion by DenisKnezBeforeNine** [fyne-io/fyne#3645] [Q&A] [answered]
Is it possible to make text selectable?

The 3 types of text that I'm aware of don't provide this functionality

	func main() {
		a := app.New()
		w := a.NewWindow("Test")
		w.CenterOnScreen()
		w.Resize(fyne.NewSize(200, 200))

		richText := widget.NewRichTextWithText("rich text")
		canvasText := canvas.NewText("canvas text", color.White)
		labelText := widget.NewLabel("label text")

		textContainer := container.NewVBox(richText, canvasText, labelText)

		w.SetContent(textContainer)
		w.ShowAndRun()
	}

<br>

This is a really useful functionality where there is human generate content. For instance I'm currently making a chat application and when a user types something the other participants in the chat maybe want to copy that text.

<br>
<br>

This is how copying text looks in slack would it be possible to implement that in fyne right now

<img width="815" alt="msg" src="https://user-images.githubusercontent.com/119591262/217756601-9561a9fd-7c3a-4982-b7cf-8ece502e3b8c.png">

**Comment by andydotxyz** [comment]
The only interactive text is Entry. We want widgets to indicate interactivity and Label doesn't encourage or indicate that action.

This is not the default on GUI apps, it is only default on web based apps (and hybrids like Electron used in Slack).

Happy to explore this further - but so far we've not seen a good solution.

**Reply by alx3dev** [reply]
It is possible to extend entry widget, so it doesn't have a cursor. Set entry function `OnChange` to always set text and refresh it. That way users can't edit entry, but can select. Dirty but it works.

**Reply by andydotxyz** [reply]
Wouldn't it be easier to just set it as Disabled? That way there is clarity to the user that it cannot be edited...

**Reply by alx3dev** [reply]
Yes, now I see disabled widgets allow selecting. Then color is another problem, unless we can change disabled color on that widget only.

**Reply by andydotxyz** [reply]
Not with the default widgets - we want to show the user that it is disabled.

**Reply by boosh** [reply]
I have a similar use case. I'm writing an app that calls the chatgpt API, and similar to the web I want to allow users to only select part of the responses. I can't use an Entry since I'm using NewRichTextFromMarkdown. Is it possible to allow users to select text in a RichText widget?

**Reply by andydotxyz** [reply]
There is no rich text interactions. After we build a RichTextEntry this may become possible.

**Reply by boosh** [reply]
Oh OK. Is there an ETA for that? Great library BTW, thanks

**Reply by andydotxyz** [reply]
The RichTextEntry was proposed for the release after the one we are working on - late 2023 probably https://github.com/fyne-io/fyne/milestone/21

**Comment by hkparker** [comment]
I'm also looking for a solution here at some point, since I'm also building a chat app and it's important to me that users can select + copy message text out of chat bubbles.  If there isn't something that makes it into the API by the time I'm looking to add this I'll probably end up making some weird fork of entry that removes all other functionality and keeps selection, but I'm a ways away from this being a priority.

**Reply by andydotxyz** [reply]
On mobile there is the secondary-tap menu with a copy option - at least that model works in the meantime...

**Comment by dweymouth** [comment, marked as answer]
Just noting here that selectable Label is arriving in 2.6.0

**Comment by limafresh** [comment]
I created a customizable message bubble widget with selection support ➡️ https://github.com/limafresh/messagebubble

In case it's still relevant.

**Reply by hkparker** [reply]
This is cool!  My message bubble currently supports hyperlinks via rich text, and I want to support the other rich text formatting features eventually, so I won't go the label route to get selectability, but it's nice to see another implementation of a bubble.
